From 33994cf93b6c31a3ab021edcea9bfd5376e94b5d Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=E5=A7=9C=E6=B5=B7=E5=A4=A9?= <1076168468@qq.com>
Date: Sat, 29 Oct 2022 16:42:34 +0800
Subject: [PATCH] =?UTF-8?q?=E4=BB=BB=E5=8A=A1=E5=AE=8C=E6=88=90=EF=BC=8C?=
 =?UTF-8?q?=E6=8F=90=E4=BA=A4=E4=BD=9C=E4=B8=9A?=
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

---
 EOSApp.c | 53 ++++++++++++++++++++++++++++++++++-------------------
 1 file changed, 34 insertions(+), 19 deletions(-)

diff --git a/EOSApp.c b/EOSApp.c
index d8f2bae..8e1788a 100644
--- a/EOSApp.c
+++ b/EOSApp.c
@@ -1,26 +1,41 @@
 #include "EOSApp.h"
 
-//
-// main 函数参数的意义：
-// argc - argv 数组的长度，大小至少为 1，argc - 1 为命令行参数的数量。
-// argv - 字符串指针数组，数组长度为命令行参数个数 + 1。其中 argv[0] 固定指向当前
-//        进程所执行的可执行文件的路径字符串，argv[1] 及其后面的指针指向各个命令行
-//        参数。
-//        例如通过命令行内容 "a:\hello.exe -a -b" 启动进程后，hello.exe 的 main 函
-//        数的参数 argc 的值为 3，argv[0] 指向字符串 "a:\hello.exe"，argv[1] 指向
-//        参数字符串 "-a"，argv[2] 指向参数字符串 "-b"。
-//
 int main(int argc, char* argv[])
 {
-	//
-	// 如果需要在调试应用程序时能够调试进入内核并显示对应的源码，
-	// 必须使用 EOS 内核项目编译生成完全版本的 SDK 文件夹，然
-	// 后使用刚刚生成的 SDK 文件夹覆盖此应用程序项目中的 SDK 文件
-	// 夹，并且 EOS 内核项目在磁盘上的位置不能改变。
-	//
-
-	/* TODO: 在此处添加自己的代码 */
-	printf("Hello world!\n");
+	//分配内存
+	int* p=VirtualAlloc(NULL,sizeof(int),MEM_RESERVE|MEM_COMMIT);
+	
+	//检查内存是否分配成功
+	if(NULL == p){
+		printf("VirtualAlloc failed, error code:0x%X\n",GetLastError());
+		return 1;
+	}
+	
+	//打印分配地址
+	printf("Allocate %d bytes virtual memory at 0x%X\n\n",sizeof(int),p);
+	
+	//打印初始值及赋值结果
+	printf("Virtual memory original value: 0x%X\n",*p);
+	*p=0xFFFFFFFF;
+	printf("Virtual memory new value: 0x%X\n",*p);
+	
+	//等待10秒
+	printf("\nWait for 10 second\n");
+	Sleep(10000);
+	
+	//释放内存
+	if(!VirtualFree(p,0,MEM_RELEASE)){
+		printf("VirtualFree failed, error code:0x%X\n",GetLastError());
+		return 2;
+	}
+	
+	printf("\nRelease virtual memory success!\n");
+	
+	//进入循环
+	printf("\nEndless loop!\n");
+	while(TRUE){
+		Sleep(0);//使线程进入就绪状态避免浪费资源
+	}
 
 	return 0;
 }
-- 
2.34.1.windows.1

